contracts/container/contract.go

```diff
@@ -491,6 +494,179 @@ func List(owner []byte) [][]byte {
 	return list
 }
 
+// AddNextEpochNodes accumulates passed nodes as container members for the next
+// epoch to be committed using [CommitContainerListUpdate]. Nodes must be
+// grouped by selector index from placement policy (SELECT clauses). Results of
+// the call operation can be received via [Nodes]. This method must be called
+// only when a container list is changed, otherwise nothing should be done.
+// Call must be signed by the Alphabet nodes.
+func AddNextEpochNodes(cID interop.Hash256, placementVector uint8, publicKeys []interop.PublicKey) {
+	if len(cID) != interop.Hash256Len {
+		panic(cst.ErrorInvalidContainerID + ": length: " + std.Itoa10(len(cID)))
+	}
+	// nolint:staticcheck
+	if placementVector >= 255 {
+		panic(cst.ErrorTooBigNumberOfNodes + ": " + std.Itoa10(int(placementVector)))
+	}
+	ctx := storage.GetContext()
+	validatePlacementIndex(ctx, cID, placementVector)
+
+	multiaddr := common.AlphabetAddress()
+	common.CheckAlphabetWitness(multiaddr)
+
+	commonPrefix := append([]byte{nextEpochNodesPrefix}, cID...)
+	commonPrefix = append(commonPrefix, placementVector)
+
+	counter := 0
+	c := storage.Find(ctx, commonPrefix, storage.RemovePrefix|storage.KeysOnly|storage.Backwards)
+	if iterator.Next(c) {
+		counter = counterFromBytes(iterator.Value(c).([]byte))
+	}
+
+	for _, publicKey := range publicKeys {
+		if len(publicKey) != interop.PublicKeyCompressedLen {
+			panic(cst.ErrorInvalidPublicKey + ": length: " + std.Itoa10(len(publicKey)))
+		}
+
+		counter++
+
+		storageKey := append(commonPrefix, counterToBytes(counter)...)
+		storage.Put(ctx, storageKey, publicKey)
+	}
+}
+
+func counterToBytes(counter int) []byte {
+	var anyCounter any = counter
+	res := anyCounter.([]byte)
+
+	switch len(res) {
+	case 0:
+		res = []byte{0, 0}
+	case 1:
+		res = []byte{res[0], 0}
+	default:
+		// only 2 bytes are expected, it should be ensured on the upper levels
+	}
+
+	// BE for correct sorting
+	first := res[0]
+	res[0] = res[1]
+	res[1] = first
+
+	return res
+}
+
+func counterFromBytes(counter []byte) int {
+	first := counter[0]
+	counter[0] = counter[1]
+	counter[1] = first
+
+	var anyCounter any = counter
+
+	return anyCounter.(int)
+}
+
+func validatePlacementIndex(ctx storage.Context, cID interop.Hash256, inx uint8) {
+	if inx == 0 {
+		return
+	}
+
+	commonPrefix := append([]byte{nextEpochNodesPrefix}, cID...)
+
+	for i := range int(inx) {
```

Comment:
I'd simplify this to checking `inx-1`, this should be sufficient.

Reply:
dont mind